File: apps/api/app/routers/keys.py
import hashlib
import secrets

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from app.auth import get_current_user_phone
from app.deps import optional_supabase

router = APIRouter(prefix="/keys", tags=["keys"])


class CreateKeyRequest(BaseModel):
    name: str
    scopes: list[str] = []


def _hash(key: str) -> str:
    return hashlib.sha256(key.encode()).hexdigest()


@router.post("")
async def create_key(
    body: CreateKeyRequest,
    phone: str = Depends(get_current_user_phone),
    db=Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    raw = f"nw_live_{secrets.token_urlsafe(32)}"
    row = db.table("nano_api_keys").insert({
        "user_phone": phone,
        "name": body.name,
        "key_hash": _hash(raw),
        "key_prefix": raw[:12],
        "scopes": body.scopes,
    }).execute()
    data = row.data[0] if row.data else {}
    return {**data, "key": raw}  # full key returned once only


@router.get("")
async def list_keys(
    phone: str = Depends(get_current_user_phone),
    db=Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    row = (
        db.table("nano_api_keys")
        .select("id,name,key_prefix,scopes,last_used_at,expires_at,is_active,created_at")
        .eq("user_phone", phone)
        .eq("is_active", True)
        .order("created_at", desc=True)
        .execute()
    )
    return row.data


@router.delete("/{key_id}")
async def revoke_key(
    key_id: str,
    phone: str = Depends(get_current_user_phone),
    db=Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    row = (
        db.table("nano_api_keys")
        .update({"is_active": False})
        .eq("id", key_id)
        .eq("user_phone", phone)
        .execute()
    )
    if not row.data:
        raise HTTPException(404, "Key not found")
    return {"revoked": True}

File: apps/worker/src/routes/keys.ts
import { Hono } from 'hono'
import type { Env } from '../index'
import { getPhone } from '../middleware/auth'
import { getSupabase } from '../lib/supabase'

const app = new Hono<{ Bindings: Env }>()

async function sha256hex(text: string): Promise<string> {
  const buf = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(text))
  return Array.from(new Uint8Array(buf)).map(b => b.toString(16).padStart(2, '0')).join('')
}

function randomToken(): string {
  const bytes = new Uint8Array(32)
  crypto.getRandomValues(bytes)
  return Array.from(bytes).map(b => b.toString(16).padStart(2, '0')).join('')
}

app.post('/', async (c) => {
  let phone: string
  try { phone = await getPhone(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const { name, scopes = [] } = await c.req.json<{ name: string; scopes?: string[] }>()
  const raw = `nw_live_${randomToken()}`
  const key_hash = await sha256hex(raw)
  const sb = getSupabase(c.env)
  const { data, error } = await sb.from('nano_api_keys').insert({
    user_phone: phone, name, key_hash, key_prefix: raw.slice(0, 12), scopes,
  }).select().single()
  if (error) return c.json({ error: error.message }, 503)
  return c.json({ ...data, key: raw })
})

app.get('/', async (c) => {
  let phone: string
  try { phone = await getPhone(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const sb = getSupabase(c.env)
  const { data } = await sb
    .from('nano_api_keys')
    .select('id,name,key_prefix,scopes,last_used_at,expires_at,is_active,created_at')
    .eq('user_phone', phone)
    .eq('is_active', true)
    .order('created_at', { ascending: false })
  return c.json(data ?? [])
})

app.delete('/:id', async (c) => {
  let phone: string
  try { phone = await getPhone(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const sb = getSupabase(c.env)
  const { data, error } = await sb
    .from('nano_api_keys')
    .update({ is_active: false })
    .eq('id', c.req.param('id'))
    .eq('user_phone', phone)
    .select()
  if (error) return c.json({ error: error.message }, 503)
  if (!data?.length) return c.json({ error: 'Key not found' }, 404)
  return c.json({ revoked: true })
})

export default app
